### PR TITLE
Replace File Process Tracking With File Write Tracking

### LINKED ISSUE #12: Remove File Process Tracking Table
Switch to focus file tracking based on the `struct file *` instances, this better maps behavior of how files work within Linux. Also we need to try to reduce data race conditions within the generic hashtable usage.

In future work we can then implement better LRU caching to do reduce duplicate file events.

### PR DESCRIPTION
**Pull request by jrmwooldridge**
Closes #12 

This is a drop in replacement to file process tracking. The file write tracking decouples itself from process tracking and behaves similar to how file handles naturally work in the linux kernel. The file write table entries also takes data racing into account. This means duplicate entries will be harder to create. There is no direct access to a file entry's data, safe copies of parts of the data is used to get or update the entry data. Reducing the direct memory access should resolve a lot of issues. The process tracking table now longer has to be concerned about file lists and we could eventually fix other things with process tracking in the future. This technically may emit more file write events, if this causes performance issues it can be fixed with a small lru cache to check for duplicates. Major performance improvements also have been made. File events writes can be up to 50% - 300% times faster depending on scenario. With the generic hashtable improvements we should see a noticable increase with better CPU utilization on load.

Signed-off-by: Jeffery Wooldridge <jwooldridge@vmware.com>

### CODE REVIEW

**Comment by jrmwooldridge** [comment]
Preventing duplicate File Write Events and more advanced performance enhancements will be in a different PR.

**Review by zhao-cb — DISMISSED** [review]

File: src/kmod/file-write-tracking.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+struct file_write_key {
+	struct file *file;
+};
+// Add inode and device when CB_EVENT
+// can handle them on generic file events
+struct file_write_entry {
+	struct HashTableNode   link;
+	struct file_write_key  key;
+	u32		       last_tgid;
```
Comment (dheater) [inline src/kmod/file-write-tracking.h:28]: Look like tgid is always a PID. Change type to match.
```suggestion
	pid_t		       last_tgid;
```

File: src/kmod/file-write-tracking.c [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+		str = "7 ZIP";
+		break;
+	case filetypeUnknown:
+	default:
+		break;
+	}
+	return str;
+}
+
+static int _show_file_tracking_table(struct HashTbl *	   hashTblp,
```
Comment (dheater) [inline src/kmod/file-write-tracking.c:345]: Remove unused parameter `struct HashTbl * hashTblp`

**Review by dheater — COMMENTED** [review]
Looks awesome!
Just a few minor cleanup comments.

File: src/kmod/file-hooks.c [hunk 3]
```diff
@@ -187,6 +171,7 @@ static struct super_block *_sb_from_file(struct file *file)
 	return sb;
 }
 
+#if 0
```
[the remark this answers is not in the document]
Comment (jrmwooldridge) [reply src/kmod/file-hooks.c:174]: It's going to be re-enabled in the next PR

File: src/kmod/file-write-tracking.c (hunk 2 above)
Comment (jrmwooldridge) [reply src/kmod/file-write-tracking.c:345]: Pretty sure it's a required parameter for a function pointer

**Review by dheater — APPROVED** [review]

**Review by zhao-cb — APPROVED** [review]

**Pull request merged**